### ISSUE TITLE
Revisiting our namespacing and importing practices

### ISSUE THREAD

**Issue by pmcharrison** [music-computing/amads#28]
Currently we write imports like this:

```py
from musmart.pt_midi_import partitura_midi_import
from musmart.durdist1 import duration_distribution_1
```

`partitura_midi_import` is actually stored in `musmart/io/pt_midi_import`, but we have some special code in `musmart/__init__.py` that allows to write the shorter version `from musmart.pt_midi_import partitura_midi_import`.

There is some additional logic in `musmart/__init__.py` that implements lazy loading of modules such as `pt_midi_import`. When you write `import musmart`, it does not currently import all the modules within `musmart`; rather, it only imports modules when you access them.

There are some benefits of the current setup:

- We don't force the user to engage with the full hierarchy of source code organisation, which (arguably) 
   improves the user experience and defends us from problems if we decide to reorganise the directory structure later.
- Import times are reduced by not importing all modules at once.

There are some problems, however:

Users still have to learn which low-level module their desired function is stored in. For example, you can't write

```py
from musmart import partitura_midi_import
```

-- you have to write 

```py
from musmart.pt_midi_import import partitura_midi_import
```

A second problem is that the simulation of a flat package structure is imperfect. When a reader sees code like this:

```py
from musmart.pt_midi_import partitura_midi_import
```

they might reasonably assume that `pt_midi_import` is a valid top-level module and hence they should be able to write this:

```py
import musmart
musmart.pt_midi_import.partitura_midi_import
```

in the same way that one might write

```py
import scipy
scipy.stats.alpha
```

but the former does not work:

```
>>> import musmart
>>> musmart.pt_midi_import.partitura_midi_import
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'musmart' has no attribute 'pt_midi_import'
```

A third problem is that linters complain that the importing code is wrong:

<img width="673" alt="image" src="https://github.com/user-attachments/assets/135a4441-c012-4ebb-b55a-667d28a5cd98">

and, relatedly, autocomplete fails with such statements.

A general observation is that we're going against mainstream Python practice here, making up our own solution to a problem that is definitely not unique, and introducing various new problems along the way.

I would like to propose the following solution which is more in line with standard practice in Python packaging.

1. Remove the lazy import logic from `musmart/__init__.py`. The main overhead in importing comes from when we import other packages with slow import times, like partitura. We should make such imports local, i.e. only run `import partitura` within a function that actually needs partitura. If we do this then our package's import times will become minimal, even if we import every script within the package on package initialisation.

2. Personally, I would prefer not to implement the flat namespace logic at all, i.e. I am happy with people writing

```py
import musmart
musmart.io.partitura_midi_import
```
or 

```py
musmart.algorithms.complexity.lz77
```

or 

```py
musmart.representations.pitch.pc_set
```

I think that this approach is actually very readable; it helps developers know where to go to edit source code; it is what people are expecting when they work with Python packages, and it is consistent with accepted best practices; it works well with autocompletes, almost like a search function (if you want to list available pitch representations, you type `musmart.pitch.` and then your IDE will list all available options; it also takes advantage of lazy importing, in that Python will only import modules you actually access).

However, I like supporting diverse approaches rather than forcing one way to do everything. How about also implementing the following option? 

```py
import musmart
from musmart.all import *

pc_set 
partitura_midi_import
```

This would be straightforward to implement using standard Python practices. Might that be the best of both worlds? People could then choose either way to write their imports, and the linters would work fine.

If we accept this way forward I am happy to do the required refactoring.

**Comment by rbdannenberg** [comment]
I see the repo has timing, rhythm and meter (3 notions of time); pitch, melody, serial (3 notions of pitch); harmony, tonality (2 notions of harmony). I thought we were going for a flatter taxonomy. Yiwen is porting some MidiToolbox functions on melody, but they include functions like tessitura and range -- should these be in amads.pitch or amads.melody? Could we have a README file in each directory to define it's purpose?

**Comment by pmcharrison** [comment]
Sorry @rbdannenberg, I had neglected to implement the reorganisation agreed in our meeting above. I have done this now, and made a few additional consolidations as suggested by you (e.g. subsuming tonality and serial into pitch). I think melody cannot be subsumed into pitch though (because a melody also has the temporal dimension). I might be persuaded to combined harmony with pitch though (so we have e.g. amads.pitch.consonance). I have also added some docstrings to the modules to explain the definitions of each module.

Personally, I think it makes sense to put those melody algorithms from MidiToolbox into our melody module. While in principle they could be applied to a polyphonic note matrix, I think in practice the idea of range and tessitura is most applicable to melodies.
